keps/sig-cloud-provider/20190422-cloud-controller-manager-migration.md

```diff
@@ -157,53 +159,35 @@ will prevent any of the v1.18 CCMs from claiming the lock. When the current hold
 
 ### Implementation Details/Notes/Constraints [optional]
 
-TODO after KEP summary and proposal is approved
+The cloud-provider needs to would need to implement an interface, that for a given controller returns the version of migration
```

Comment:
> that for a given controller returns the version of migration

Help me understand this interface--who is the intended consumer?  If its the controller code, why not have the cloud provider code, given a controller returns the name of the lock it should use, since that's what the controller wants to figure out anyway?

Reply:
I think my only concern with this idea is that I would like us to have it work in processes which don't import the cloud provider. It would be nice if we could continue using this mechanism to migrate late moving controllers out of the KCM, even after cloud provider is gone from the KCM.

Reply:
cc @andrewsykim

Reply:
> Help me understand this interface--who is the intended consumer? If its the controller code, why not have the cloud provider code, given a controller, return the name of the lock it should use, since that's what the controller wants to figure out anyway?

Agree with @nckturner on this - I think the migration mechanism needs to be generic, but the entry point and what determines the lock can be use-case specific.

Reply:
would this mean that we would add provider-specific implementation (maybe somewhere in legacy-providers package). Honestly I'm fine with it too, as long as the code-org subproject is aware of it.

Also thinking out-loud: If it seems that everyone wants to migrate everything at once, or do the network-controllers then the rest, we could get out with 2 implementations.

defering to @andrewsykim @cheftako

Reply:
> I think my only concern with this idea is that I would like us to have it work in processes which don't import the cloud provider.

Agreed, it should work in cases where the cloud provider is gone.  I think this is possible.

For example, there is an interface in component-base/client-go (or something similar) and the implementation for the initial migration is in cloud provider.  So whenever KCM wants to do leader things, it first has to call:

```
// very pseudo code 
lock := cloud.GetLock("1.17", "node-controller", "kcm")

// in cloud provider
func (c *Cloud) GetLock(k8sVersion, controllerName, componentName string) {
    // impl returns lock name
}
```

Then when the cloud provider is deleted, the call to the above function must be modified slightly to use a similar function implemented in the KCM. 

> I think the migration mechanism needs to be generic, but the entry point and what determines the lock can be use-case specific.

Is that at all similar to what you were thinking @andrewsykim?